tests/DCPS/Deadline/DataReaderListenerImpl.cpp

```diff
@@ -9,149 +10,142 @@
 using namespace Messenger;
 
 DataReaderListenerImpl::DataReaderListenerImpl()
-  : mutex_()
-  , matched_condition_(mutex_)
+  : matched_mutex_()
+  , matched_cv_(matched_mutex_)
   , matched_(0)
-  , num_arrived_(0)
-  , requested_deadline_total_count_ (0)
+  , received_mutex_()
+  , received_cv_(received_mutex_)
+  , received_(0)
+  , count_mutex_()
+  , requested_deadline_total_count_(0)
 {
 }
 
 DataReaderListenerImpl::~DataReaderListenerImpl()
 {
 }
 
-void
-DataReaderListenerImpl::on_data_available(DDS::DataReader_ptr)
+bool DataReaderListenerImpl::wait_matched(int count, const OpenDDS::DCPS::TimeDuration& max_wait) const
 {
-  ACE_GUARD(ACE_Thread_Mutex, guard, mutex_);
-  ++num_arrived_;
-  ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) DataReaderListenerImpl::on_data_available\n")));
+  using namespace OpenDDS::DCPS;
+  Lock lock(matched_mutex_);
+  if (!lock.locked()) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: DataReaderListenerImpl::wait_matched: failed to lock\n"));
+    return false;
+  }
+  const MonotonicTimePoint deadline = MonotonicTimePoint::now() + max_wait;
+  while (count != matched_) {
+    switch (matched_cv_.wait_until(deadline)) {
+    case CvStatus_NoTimeout:
+      ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) DataReaderListenerImpl::wait_matched: %d\n"), matched_));
+      break;
+
+    case CvStatus_Timeout:
+      ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: DataReaderListenerImpl::wait_matched: Timeout\n"));
+      return false;
+
+    case CvStatus_Error:
+      ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: DataReaderListenerImpl::wait_matched: "
+        "Error in wait_until\n"));
+      return false;
+    }
+  }
+  return true;
+}
+
+void DataReaderListenerImpl::on_subscription_matched(DDS::DataReader_ptr, const DDS::SubscriptionMatchedStatus& status)
+{
+  Lock lock(matched_mutex_);
```

Comment:
In general, I'd like to avoid macros if possible. In this particular case, the ACE_GUARD macro was avoided because of the following comment.
 
OpenDDS/ACE_wrappers/ace/Global_Macros.h (Lines 137-151):
 * @warning
 *   Use of ACE_GUARD() is rarely correct.  ACE_GUARD() causes the
 *   current function to return if the lock is not acquired.  Since
 *   merely returning (no value) almost certainly fails to handle the
 *   acquisition failure and almost certainly fails to communicate the
 *   failure to the caller for the caller to handle, ACE_GUARD() is
 *   almost always the wrong thing to do.  The same goes for
 *   ACE_WRITE_GUARD() and ACE_READ_GUARD() .  ACE_GUARD_REACTION() is
 *   better because it lets you specify error handling code.
 */
#if !defined (ACE_GUARD_ACTION)
#define ACE_GUARD_ACTION(MUTEX, OBJ, LOCK, ACTION, REACTION) \
   ACE_Guard< MUTEX > OBJ (LOCK); \
   if (OBJ.locked () != 0) { ACTION; } \   //?? Line 149
   else { REACTION; }    //?? Line 151

However, I do see your valid point that the lock may fail. Since the Lock here is an ACE_Guard (in .h: typedef ACE_Guard<Mutex> Lock;), I added the following code - exactly the same logic as the macro ACE_GUARD_ACTION (Line 149-151) does, and changed the function return value from void to int and also the caller code.

  Lock lock(received_mutex_);
  if (!lock.locked()) {
    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) DataReaderListenerImpl::wait_all_received failed to get lock.\n")));
    return 1;
  }

Reply:
I've seen that comment before and planned to see if there was something to be done about it when we start to move away from ACE.

In the mean time I would say that it depends on how critical a failure would be where the `ACE_GUARD` is. Here in tests it's not critical. Also compared against the work of writing an error log statement, generally I think just returning is fine and would be at least easier to debug than locking and not checking at all.

Reply:
Yeah, it would be nice if we could use std library objects to replace ACE macros. I tried that in another test but only partially succeeded, since I had to use #ifdef ACE_HAS_CPP11, otherwise it wont work for pre-c++11 configurations we support. That's why I like to use typedef's in places like this. 

We can easily change to std objects by replacing:
typedef ACE_SYNCH_MUTEX Mutex;
typedef ACE_Guard<Mutex> Lock;

with:
typedef std::mutex Mutex;
typedef std::unique_lock<Mutex> Lock;

So the code replacing ACE macro with an explicit call to lock.locked() and an error message is worth it. It's also consistent in this test with the publisher side: 
bool DataWriterListenerImpl::wait_matched(long count, const OpenDDS::DCPS::TimeDuration& max_wait) const
{
  using namespace OpenDDS::DCPS;
  Lock lock(mutex_);
  if (!lock.locked()) {
    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: DataWriterListenerImpl::wait_matched: failed to lock\n"));
    return false;
  }

Reply:
Okay that's probably fine, but regardless of what you do all the locks should be checked.